### PR TITLE
Add support for new braille display from Nattiq

### PR DESCRIPTION
**Pull request by m-noman**
### Change log entry:

Section: New features, Changes, Bug fixes

Adding support for the new Braille display from Nattiq Technologies
http://www.nattiq.com/en/node/1694

### CODE REVIEW

**Comment by m-noman** [comment]
After further checking and testing, it seems the driver is stable and good to merge, 
so that NVDA can officially have the support for the Nattiq nBraille systems.
@leonardder @feerrenrut @Adriani90 
if someone can review,
Thanks.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]:
```suggestion
			self._serial.write(b"reset")

```

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Could you actually store this in a constant? Please use raw byte strings here as well.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: This is quite a long time.

```suggestion
		self._serial.waitForRead(self.timeout * 10)

```

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Rather than a constant, please use the timeout property on the driver. See the handy tech driver for an example.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Please use constants that describe what the commands do.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Could you move this upwards, so we have 0, 1, 2, 3, 4, etc. in sequence?

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Is this obsolete?

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: I think this should be a byte string:

```suggestion
		cells = b"-".join(cell.to_bytes(1, "little") for cell in cells)

```

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Again, please use constants here. You could also make this a dictionary that maps constants to ids.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: When do you expect this to happen?

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Overall, this looks pretty nice.

Could you say anything about how widely this display is used?

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: This is done and tested without problems.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: This and all other constants are now ids at the start.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: I edited this out, and it seems to works fine, so I will keep it.
The reason why I put 3 is that I was not sure how long the time is needed for the system unit to do the "reset", but it seems fast enough that the change to "self.timeout * 10" does not create a problem.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: This is done for all timeout mentions

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: I found this in many of the braille drivers, so I had it there just in cases, I removed it and there are no changes in the braille testing, so I guess it is not needed anymore, even though some drivers in NVDA are still having it.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: This change seemed to break the output, not sure why, it might be that the cells data does not need to be in bytes format ( I think they are already though? ) so that only the dash "-" is now encoded, and with that the braille display works fine.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: I removed this. 
Not sure why I put it in the first place, I had some problems at first about valueError,
either it had something to do with the python libraries and so when I updated now it does not show up anymore, or for some reasons I was getting this error ( corrupt data or some serial problem )
I tested now again and will keep doing tests, so far it seems OK without this and all router keys works fine.

**Comment by m-noman** [comment]
@leonardder
Thank you for the feedback,
This Braille display is actually New.
It was released around December last year, so normally it is rather new and does not yet have a huge user-base, but I expect it may get more exposure due to its low price comparing to other brands.
It also have support for other screen readers like HAL/Supernova, and so this is how I created this driver. All testing I am doing is on the braille display unit.

I edited out the comments, I am not sure if it is now better, or need more comments/explanations?
I will keep testing and will await the feedback for the changes.

Thank you.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: As you say this is a new driver, you can change this as follows:

```suggestion
# Copyright (C) 2020 NV Access Limited, <your full name>

```

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Typo

```suggestion
# Initialization response id

```

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Why would you wait for input after pressing this key?

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Same as above

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Same as above

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Same as above

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: Are you really sure that this works? You're mixing str with bytes here. Please see my earlier suggestion.

**Comment by leonardder** [comment]
Are there actually any docs about this display, i.e. this driver is implemented using serial communication, but is this a usb to serial converter? If so, you might want to add the vid and pid of the converter to bdDetect in order for automatic braille display detection to work.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: oh, I edited it out, thanks.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: ok, done.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: sorry, I forgot to remove that, it was just there for some debugging on the receiving side.
I removed them for all buttons.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: I edited out to your suggestion, I will test the build again,
from what I understood, that it does not matter if NVDA sent the cells data as bytes or as a string, because whether they were this or that, they will be converted to integers for the Braille display. so mostly it should not matter, but I need to test again with the unit I have and see how it performs

**Comment by m-noman** [comment]
> Are there actually any docs about this display, i.e. this driver is implemented using serial communication, but is this a usb to serial converter? If so, you might want to add the vid and pid of the converter to bdDetect in order for automatic braille display detection to work.

I don't think there are things that are publicly available.
but the Braille display is not using a usb to serial converter, rather it is using a hardware based usb controller. It is based on Atmel micro-controllers. Though its PID/VID will be kind of generic as they are for the Atmel micro-chip, but I checked the bdDetect.py file and it is possible to add them there for auto-detection. Though manually, the braille display seems to works fine, on starts/reboots/ and even re-installs ( not sure how? or this might be normal )

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: Nope, with converting the cells data to bytes, it does not works. It gets detected fine, and input is OK, but no output. So it seems it requires the cells data (0-255) to be in String format.
Do you suggest to keep this 
b"-".join(str(cell) for cell in cells)
or to keep the dash in string?
"-".join(str(cell) for cell in cells)

Both works fine after testings, because the dash "-" is only a separator and it gets ignored. As the important part is the cells data to display them for each cell.
Thanks.

**Comment by m-noman** [comment]
@leonardder 
I just noticed that now NVDA uses python 3 instead of 2.
I guess that was the reason why it worked for me with either bytes or strings.
Now when I updated to python 3 and VS2019, I can not write strings like before,
and because the unit expect a string encoded value, so I will just use the dash and cell data as strings then encode the output for serial.write().
I tried it out and it seems to works fine. I will await your review.
Also the auto-detect works now.

Thank you.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [reply source/brailleDisplayDrivers/nattiqbraille.py]: How important are these dashes? Are they prescribed by the protocol?

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: I"m sorry, I meant the < and > not to appear in the result.

```suggestion
# Copyright (C) 2020 NV Access Limited, Mohammed Noman

```

File: source/bdDetect.py
Comment (leonardder) [inline source/bdDetect.py]: Could you please position the block above # superBrl, so the entries will stay in alphabetical order?

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/nattiqbraille.py]: In Python 3, there is a strict separation between strings and bytes objects. Strings assume you're working with textual data, which is not the case when writing cells to a braille display, so my personal  preference would be avoiding encoding like this.

If the dashes are really important, you can do:

`b"-".join(cell.to_bytes(1, "little", signed=False) for cell in cells)`

If the dashes can be ommited, you can just use `bytes(cells)` to convert the list of integers to bytes.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: > How important are these dashes? Are they prescribed by the protocol?

Yes, they are separating the data for the cells. The unit receives the whole line one time instead of feeding it one by one, so it needs to have a separator of the cell data.
Then after the unit receive the line data, it converts each cell string to its byte value (8-bit cells for 8-dots) which means the (0-255) gets into each cell by the order it gets from NVDA.

File: source/bdDetect.py
Comment (m-noman) [reply source/bdDetect.py]: ok done.

File: source/brailleDisplayDrivers/nattiqbraille.py
Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: It is still messing something up, because when I use the cells as byte and then send them, for some reasons the unit detects nothing at all, it just put the whole braille line as clear or empty.
While when using the cells as string data, then encoding them for the writing, the braille works fine on the unit. 
it might be something to do with how the unit interrupts the data itself.

Comment (leonardder) [reply source/brailleDisplayDrivers/nattiqbraille.py]: Ugh, it's because my assumptions where wrong. Of course, str(x) just returns a string with x, not the byte value of x. That also clarifies the dashes. Unfortunately, bytes(x) returns a bytes object length x, not a bytes object containing x.

I think I'd leave is as is, but may be @feerrenrut has additional ideas.

Comment (m-noman) [reply source/brailleDisplayDrivers/nattiqbraille.py]: rsz_a_20200218_112722
So far, everything is working as it should with input and output.
It will be great to have more ideas about the optimal code, so that hopefully this driver works perfectly and does not create problems in the future
@leonardder  Thank you for the review.

**Review by LeonarddeR — COMMENTED** [review]
I think the code is nice as it is now, thanks for your contribution.

Next steps would be updating the user guide

* Add a paragraph about the display to chapter 15 (supported braille displays) of the user guide: userDocs\en\userGuide.t2t
* Add the driver to the `Displays supporting automatic detection in the background` section

**Comment by m-noman** [comment]
> 
> 
> I think the code is nice as it is now, thanks for your contribution.
> 
> Next steps would be updating the user guide
> 
>     * Add a paragraph about the display to chapter 15 (supported braille displays) of the user guide: userDocs\en\userGuide.t2t
> 
>     * Add the driver to the `Displays supporting automatic detection in the background` section

I added the display to the list of automatic supported displays, and also a small paragraph with the keys for the display.
I put it above BRLTTY, I do not think this needs to be ordered by alphabetic? because the others are not in order. I only choose it to be above BRLTTY, as that is not a display but rather a protocol/program, so technically, it is now the last ( if the order was meant to be from older to newer )
@leonardder  if there are any problems, or it needs to have more information, I will try to add more. but mostly I do not think we need to have more than this, as the link I attached ( in the very first post ) have all the technical details etc. 

Thanks.

File: user_docs/en/userGuide.t2t
Comment (leonardder) [inline user_docs/en/userGuide.t2t]:
```suggestion
Windows 10 detects the Braille Displays once connected, you may need to install USB drivers if using older versions of Windows (below Win10).
You can get them from the manufacturer's website.
```

File: user_docs/en/userGuide.t2t
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: Could you add the previous/next line assignments?

**Review by LeonarddeR — COMMENTED** [review]
Three small things:

* Please start a new line after every sentence in the user guide
* manufacturer website > manufaturer's website
* Also, see my comment about missing assignments.

File: user_docs/en/userGuide.t2t
Comment (m-noman) [reply user_docs/en/userGuide.t2t]: ok, done.

File: user_docs/en/userGuide.t2t
Comment (m-noman) [reply user_docs/en/userGuide.t2t]: ops, sorry, I added them now.

Comment (m-noman) [reply user_docs/en/userGuide.t2t]: @leonardder 
if there is anything else that seems missing, tell me so that I edit it out.
Thanks again.

**Review by LeonarddeR — APPROVED** [review]

**Review by feerrenrut — APPROVED** [review]
This looks fine to me, though I am unable to test it. I will wait for a review from @michaelDCurran

The release process has already started for 2020.1 I propose this goes into the 2020.2 release.

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
@feerrenrut  to be clear, obviously I can't test this either as I don't have access to one of these displays.

**Pull request merged**
